pkg/services/meta/containers.go

```diff
@@ -587,3 +597,115 @@ func (s *containerStorage) handleNewEpoch(e uint64) error {
 
 	return errors.Join(mptErr, dbErr)
 }
+
+func expandChildren(st storage.Store, rootOIDs []byte) []byte {
+	if len(rootOIDs) == 0 {
+		return nil
+	}
+
+	// sorting before every SEEK is used for exact single operation for every
+	// children searching subroutine; search is done in 3 steps:
+	//  1. root -> last/link object search
+	//  2. last/link -> first part ID search
+	//  3. first part ID -> all children search
+
+	var childrenWithParentID [][]byte
+	var rng storage.SeekRange
+	rng.Prefix = slices.Concat([]byte{attrPlainToOIDIndex}, []byte(objectsdk.FilterParentID), object.AttributeDelimiter)
+
+	rootsSorted := slices.SortedFunc(slices.Chunk(rootOIDs, oid.Size), bytes.Compare)
+	rng.Start = rootsSorted[0]
+
+	st.Seek(rng, func(k, _ []byte) bool {
+		if len(k) < len(rng.Prefix)+oid.Size+attributeDelimiterLen+oid.Size {
+			// unexpected, just skip
+			return true
+		}
+		currRoot := k[len(rng.Prefix) : len(rng.Prefix)+oid.Size]
+		for len(rootsSorted) > 0 {
+			switch bytes.Compare(currRoot, rootsSorted[0]) {
+			case -1:
+				return true
+			case +1:
+				rootsSorted = rootsSorted[1:]
+				continue
+			case 0:
+				childrenWithParentID = append(childrenWithParentID, slices.Clone(k[len(rng.Prefix)+oid.Size+attributeDelimiterLen:]))
+				rootsSorted = rootsSorted[1:]
+				return true
+			}
+		}
+		if len(rootsSorted) == 0 {
+			return false
+		}
+
+		return true
+	})
+
+	if len(childrenWithParentID) == 0 {
+		// all objects are roots, no additional children
+		return rootOIDs
+	}
+
+	firstPartOIDs := make([][]byte, 0, len(childrenWithParentID))
+	slices.SortFunc(childrenWithParentID, bytes.Compare)
+	rng.Prefix = []byte{oidToAttrIndex}
+	rng.Start = slices.Concat(childrenWithParentID[0], []byte(objectsdk.FilterFirstSplitObject), object.AttributeDelimiter)
+	st.Seek(rng, func(k, _ []byte) bool {
+		currChild := k[1 : 1+oid.Size]
```

Comment:
https://github.com/nspcc-dev/neofs-node/pull/3257#discussion_r2016321087

Reply:
can't promise i wouldn't suggest such code again; i do not feel these checks are useful. fixed